File: fbcore/src/main/java/com/facebook/common/util/TriState.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.common.util;

import com.facebook.infer.annotation.Functional;
import com.facebook.infer.annotation.Nullsafe;
import javax.annotation.Nullable;

/** Generic tri-state enum for boolean values that can also be unset. */
@Nullsafe(Nullsafe.Mode.STRICT)
public enum TriState {
  YES,
  NO,
  UNSET,
  ;

  /** @return whether this value is set; that is, whether it is YES or NO. */
  @Functional
  public boolean isSet() {
    return this != UNSET;
  }

  /**
   * Returns the value of the {@link TriState} enum that corresponds to the specified {@code
   * boolean}.
   *
   * <p>This method deliberately declares {@code boolean} as its param type rather than {@link
   * Boolean} because:
   *
   * <ol>
   *   <li>Declaring {@link Boolean} would likely result in a bunch of unnecessary autoboxing.
   *   <li>Anyone who finds himself using a {@link Boolean} instead of a {@code boolean} for its
   *       nullability should replace the {@link Boolean} with a {@link TriState}, anyway.
   * </ol>
   */
  @Functional
  public static TriState valueOf(boolean bool) {
    return bool ? YES : NO;
  }

  @Functional
  public static TriState valueOf(@Nullable Boolean bool) {
    return bool != null ? valueOf(bool.booleanValue()) : TriState.UNSET;
  }

  /**
   * Returns the {@code boolean} value that corresponds to this {@link TriState}, if appropriate.
   *
   * @return {@code true} if {@code this} is {@link TriState#YES} or {@code false} if {@code this}
   *     is {@link TriState#NO}
   * @throws IllegalStateException if {@code this} is {@link TriState#UNSET}.
   */
  @Functional
  public boolean asBoolean() {
    switch (this) {
      case YES:
        return true;
      case NO:
        return false;
      case UNSET:
        throw new IllegalStateException("No boolean equivalent for UNSET");
      default:
        throw new IllegalStateException("Unrecognized TriState value: " + this);
    }
  }

  /**
   * Returns the {@code boolean} value that corresponds to this {@link TriState}, if appropriate.
   *
   * @param defaultValue default value to use if not set
   * @return {@code true} if {@code this} is {@link TriState#YES} or {@code false} if {@code this}
   *     is {@link TriState#NO} or {@code defaultValue} if {@code this} is {@link TriState#UNSET}.
   */
  @Functional
  public boolean asBoolean(boolean defaultValue) {
    switch (this) {
      case YES:
        return true;
      case NO:
        return false;
      case UNSET:
        return defaultValue;
      default:
        throw new IllegalStateException("Unrecognized TriState value: " + this);
    }
  }

  /**
   * Returns the {@code Boolean} value that corresponds to this {@link TriState}, if appropriate.
   *
   * @return {@link Boolean#TRUE} if {@code this} is {@link TriState#YES} or {@link Boolean#FALSE}
   *     if {@code this} is {@link TriState#NO} or {@code null} if {@code this} is {@link
   *     TriState#UNSET}.
   */
  @Functional
  public @Nullable Boolean asBooleanObject() {
    switch (this) {
      case YES:
        return Boolean.TRUE;
      case NO:
        return Boolean.FALSE;
      case UNSET:
        return null;
      default:
        throw new IllegalStateException("Unrecognized TriState value: " + this);
    }
  }

  @Functional
  public int getDbValue() {
    switch (this) {
      case YES:
        return 1;
      case NO:
        return 2;
      case UNSET:
      default:
        return 3;
    }
  }

  @Functional
  public static TriState fromDbValue(int value) {
    switch (value) {
      case 1:
        return YES;
      case 2:
        return NO;
      case 3:
      default:
        return UNSET;
    }
  }
}

File: fbcore/src/main/java/com/facebook/common/util/TriState.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.common.util

import com.facebook.infer.annotation.Functional

/** Generic tri-state enum for boolean values that can also be unset. */
enum class TriState {

  YES,
  NO,
  UNSET,
  ;

  @get:Functional
  val isSet: Boolean
    /** @return whether this value is set; that is, whether it is YES or NO. */
    get() = this != UNSET

  /**
   * Returns the `boolean` value that corresponds to this [TriState], if appropriate.
   *
   * @return `true` if `this` is [TriState#YES] or `false` if `this` is [TriState#NO]
   * @throws IllegalStateException if `this` is [TriState#UNSET].
   */
  @Functional
  fun asBoolean(): Boolean =
      when (this) {
        YES -> true
        NO -> false
        UNSET -> throw IllegalStateException("No boolean equivalent for UNSET")
        else -> throw IllegalStateException("Unrecognized TriState value: $this")
      }

  /**
   * Returns the `boolean` value that corresponds to this [TriState], if appropriate.
   *
   * @param defaultValue default value to use if not set
   * @return `true` if `this` is [TriState#YES] or `false` if `this` is [TriState#NO] or
   *   `defaultValue` if `this` is [TriState#UNSET].
   */
  @Functional
  fun asBoolean(defaultValue: Boolean): Boolean =
      when (this) {
        YES -> true
        NO -> false
        UNSET -> defaultValue
        else -> throw IllegalStateException("Unrecognized TriState value: $this")
      }

  /**
   * Returns the `Boolean` value that corresponds to this [TriState], if appropriate.
   *
   * @return [Boolean#TRUE] if `this` is [TriState#YES] or [Boolean#FALSE] if `this` is
   *   [TriState#NO] or `null` if `this` is [TriState#UNSET].
   */
  @Functional
  fun asBooleanObject(): Boolean? =
      when (this) {
        YES -> true
        NO -> false
        UNSET -> null
        else -> throw IllegalStateException("Unrecognized TriState value: $this")
      }

  @get:Functional
  val dbValue: Int
    get() {
      return when (this) {
        YES -> 1
        NO -> 2
        UNSET -> 3
        else -> 3
      }
    }

  companion object {
    /**
     * Returns the value of the [TriState] enum that corresponds to the specified `boolean`.
     *
     * This method deliberately declares `boolean` as its param type rather than [Boolean] because:
     *
     * 1. Declaring [Boolean] would likely result in a bunch of unnecessary autoboxing.
     * 1. Anyone who finds himself using a [Boolean] instead of a `boolean` for its nullability
     *    should replace the [Boolean] with a [TriState], anyway.
     */
    @JvmStatic
    @Functional
    fun valueOf(bool: Boolean): TriState = if (bool) TriState.YES else TriState.NO

    @JvmStatic
    @Functional
    fun valueOf(bool: Boolean?): TriState = if (bool != null) valueOf(bool) else TriState.UNSET

    @JvmStatic
    @Functional
    fun fromDbValue(value: Int): TriState =
        when (value) {
          1 -> TriState.YES
          2 -> TriState.NO
          3 -> TriState.UNSET
          else -> TriState.UNSET
        }
  }
}
